File: inc.graphql-resources/backend/kotlin/config/ObjectScalar.kt
package com.planetbiru.graphqlapplication.config

import graphql.language.Value
import graphql.language.ObjectValue
import graphql.schema.Coercing
import graphql.schema.CoercingParseLiteralException
import graphql.GraphQLContext
import org.springframework.stereotype.Component
import java.util.Locale

/**
 * Implements the coercing logic for a custom GraphQL "Object" scalar.
 * This class defines how to serialize, parse values, and parse literals for a generic object type,
 * allowing arbitrary JSON-like objects to be used in the GraphQL schema.
 */
@Component
class ObjectScalar : Coercing<Any?, Any?> {
    /**
     * Serializes a data fetcher result into a type that can be sent to the client.
     * For this scalar, it performs a pass-through, returning the object as-is.
     *
     * @param dataFetcherResult The object to serialize.
     * @param graphQLContext The context for the GraphQL execution.
     * @param locale The locale for the serialization.
     * @return The serialized object.
     */
    override fun serialize(dataFetcherResult: Any, graphQLContext: GraphQLContext, locale: Locale): Any? {
        return dataFetcherResult
    }

    /**
     * Parses a value from a GraphQL variable.
     * For this scalar, it performs a pass-through, returning the input value as-is.
     *
     * @param input The value from the GraphQL variables.
     * @param graphQLContext The context for the GraphQL execution.
     * @param locale The locale for parsing.
     * @return The parsed value.
     */
    override fun parseValue(input: Any, graphQLContext: GraphQLContext, locale: Locale): Any? {
        return input
    }

    /**
     * Parses a value from a literal in the GraphQL query string.
     * It expects an `ObjectValue` and converts it into a map of key-value pairs.
     *
     * @param input The GraphQL literal value from the query.
     * @param variables The variables provided with the query.
     * @param graphQLContext The context for the GraphQL execution.
     * @param locale The locale for parsing.
     * @return A map representing the object, or throws [CoercingParseLiteralException] if the input is not an `ObjectValue`.
     */
    fun parseLiteral(
        input: Value<*>,
        variables: Map<String, Any>,
        graphQLContext: GraphQLContext,
        locale: Locale
    ): Any? {
        return if (input is ObjectValue) {
            input.objectFields.associate { it.name to it.value }
        } else {
            throw CoercingParseLiteralException("Expected an ObjectValue literal but was: $input")
        }
    }
}

File: inc.graphql-resources/backend/php/inc/ObjectScalar.php
<?php

use GraphQL\Language\AST\Node;
use GraphQL\Language\AST\BooleanValueNode;
use GraphQL\Language\AST\FloatValueNode;
use GraphQL\Language\AST\IntValueNode;
use GraphQL\Language\AST\StringValueNode;
use GraphQL\Type\Definition\ScalarType;

/**
 * Class ObjectScalar
 *
 * A custom GraphQL scalar type that represents any JSON-like primitive value.
 *
 * This scalar allows GraphQL fields to accept and return flexible data types
 * such as:
 * - string
 * - integer
 * - float
 * - boolean
 * - null
 *
 * It is useful for situations where the schema requires supporting dynamic
 * values without enforcing strict type definitions. Complex structures such as
 * lists and objects can be supported with additional parsing logic if required.
 */
class ObjectScalar extends ScalarType
{
    /** @var string The name of the custom scalar. */
    public $name = 'Object';

    /** @var string Description of the scalar behavior and purpose. */
    public $description = 'A custom scalar that can represent any JSON-like value: string, int, float, boolean, or null.';

    /**
     * Serializes internal PHP values to be sent to the client.
     *
     * This method returns the value unchanged, allowing any primitive
     * JSON-compatible type to pass through.
     *
     * @param mixed $value The PHP value to serialize.
     * @return mixed The serialized value sent back to the client.
     */
    public function serialize($value)
    {
        return $value;
    }

    /**
     * Parses externally provided variable values into the internal
     * representation used by the application.
     *
     * This implementation returns the value without modification.
     *
     * @param mixed $value The input value from GraphQL variables.
     * @return mixed The parsed value.
     */
    public function parseValue($value)
    {
        return $value;
    }

    /**
     * Parses GraphQL literal values (inline values in queries)
     * into PHP types.
     *
     * Supported types include:
     * - StringValueNode → string
     * - IntValueNode → integer
     * - FloatValueNode → float
     * - BooleanValueNode → boolean
     * - NullValueNode → null (implicitly handled)
     *
     * Parsing of complex types such as lists and objects is not implemented
     * here but can be added if needed.
     *
     * @param Node $valueNode The AST node representing the literal value.
     * @param array|null $variables Variables passed alongside the query (unused).
     * @return mixed The converted PHP value.
     */
    public function parseLiteral(Node $valueNode, ?array $variables = null)
    {
        $result = null;

        if ($valueNode instanceof StringValueNode) {
            $result = $valueNode->value;
        } elseif ($valueNode instanceof IntValueNode) {
            $result = (int) $valueNode->value;
        } elseif ($valueNode instanceof FloatValueNode) {
            $result = (float) $valueNode->value;
        } elseif ($valueNode instanceof BooleanValueNode) {
            $result = (bool) $valueNode->value;
        }
        // For NullValueNode, $result remains null.
        // Parsing ListValueNode or ObjectValueNode would require additional logic.

        return $result;
    }
}
